core/domain/expression_parser.py

```diff
@@ -120,6 +147,7 @@ def tokenize(expression):
         Exception: Invalid token.
     """
     expression = expression.replace(' ', '')
+    expression = replace_abs_symbol_with_text(expression)
```

Comment:
> I don't get it -- why are you doing this? You should just be writing tests to assert that usage of those symbols results in an invalid expression/equation, not trying to accept expressions/equations with it. (As you say, it isn't an expected input from the frontend, right?)

Oh, I misunderstood. I'll remove this and just add the check then.

> The current implementation is wrong anyway because it does not handle nesting properly, e.g. |1 - |3-5| + 4|. This is why it is important to define a fixed schema and stick to it. We have also talked before about not doing ad hoc parsing, which this is an example of.

Hm, that's interesting. Yeah, I didn't think about nested absolute signs, but how do we even handle this? I mean since there is no distinction between the opening and closing parens, how would we even pair something like this? For instance, in `||3+4|-2|a|+3|`, there can be two syntactically valid transformations: `((3+4)-2(a)+3)` and `((3+4)-2)a(+3)`. How would we decide which one's the intended one? Also, just a general question -- would we really see a lot of examples with nested abs signs in practice?

> Also, as a meta-comment -- please do not do this sort of thing, where a comment is given, you do something completely different (without stating that at all), and then say "Done". That's ... just not the way to go.

Sorry about that, I'll take care from now on :)

Reply:
> Hm, that's interesting. Yeah, I didn't think about nested absolute signs, but how do we even handle this? I mean since there is no distinction between the opening and closing parens, how would we even pair something like this? For instance, in ||3+4|-2|a|+3|, there can be two syntactically valid transformations: ((3+4)-2(a)+3) and ((3+4)-2)a(+3). How would we decide which one's the intended one? 

There are various ways, but the easiest is to define validity such that the accepted valid formats are always unambiguous. This is why it's always important to define "validity" clearly for something like this, and for any parsing problem that you do -- you'll probably need some tweaks to that definition to avoid ambiguity.

> Also, just a general question -- would we really see a lot of examples with nested abs signs in practice?

I don't know, but that's irrelevant. They do exist and your parser needs to be able to process 100% of inputs. (I have seen them in math questions before, if that is what you're asking.)